### PR TITLE
[4.x] JGRP-2487 TLS Transport

### PR DESCRIPTION
**Pull request by tristantarrant**
https://issues.redhat.com/browse/JGRP-2487

This contains some fixes to make SocketFactories that return SSL sockets work in TCP (UDP and TCP_NIO are not supported).
Additionally the GossipRouter CLI has been enhanced to allow SSL configuration, including the ability to set SNI names.
@pruivo can you please add your review.
Once approved, I'll create a forward port for 5.x

### CODE REVIEW

**Comment by belaban** [comment]
OK, I'll take a look tomorrow.

**Comment by tristantarrant** [comment]
Note that the `wildfly-elytron` dependency has been added with `test` scope. JGroups will not depend on it :)

File: ivy.xml [hunk 1]
```diff
@@ -20,6 +20,7 @@
         <dependency org="org.slf4j"                name="slf4j-api"               rev="1.7+"/>
         <dependency org="org.testng"               name="testng"                  rev="6.14.+"/>
         <dependency org="com.beust"                name="jcommander"              rev="1.+"/>
+        <dependency org="org.wildfly.security"     name="wildfly-elytron"         rev="1.16.+"/>
```
Comment (pruivo) [inline ivy.xml:23]: wildfly-commons is missing. I'm getting NoClassDefFoundError when running `ant all-tests report`
```
org/wildfly/common/Assert

java.lang.NoClassDefFoundError
at org.wildfly.security.x500.cert.SelfSignedX509CertificateAndSigningKey$Builder.setDn(SelfSignedX509CertificateAndSigningKey.java:178)
at org.jgroups.protocols.TLSTest.createSelfSignedCertificate(TLSTest.java:159)
at org.jgroups.protocols.TLSTest.init(TLSTest.java:73)
```

Comment (pruivo) [reply ivy.xml:23]: looks like `SASL*Test` tests are affected by this as well.

**Review by pruivo — APPROVED** [review]

**Review by belaban — COMMENTED** [review]
It would be nice if GossipRouter had feasible default for tls_provider etc, so if we chose TSL, we would not have to provide all of the detail configuration.

**Comment by belaban** [comment]
OK, so I've merged your changes with slight modifications. Please look at TLS/SASLTest...

**Pull request merged**